File: SuperiorPackGroup/BLL/ShiftsBLL.vb
Imports SuperiorPackGroup.SPGTableAdapters
Imports System.Text
Imports DXDAL.SPGData
Imports DevExpress.Xpo

<System.ComponentModel.DataObject()> _
Public Class ShiftsBLL

    Private m_ShiftsTableAdapter As ShiftsTableAdapter = Nothing
    Public ReadOnly Property Adapter() As ShiftsTableAdapter

        Get
            If m_ShiftsTableAdapter Is Nothing Then
                m_ShiftsTableAdapter = New ShiftsTableAdapter()
                m_ShiftsTableAdapter.Connection.ConnectionString = My.Settings.UserConnectionString
            End If
            Return m_ShiftsTableAdapter
        End Get

    End Property

    Private Sub UpdateAuditTrail(ByVal ModifiedRecord As SPG.ShiftsRow, ByVal originalRecord As Object())

        Dim builder As New StringBuilder(String.Empty)
        Dim recordFields As Integer = (ModifiedRecord.ItemArray.Length - 1)
        Dim i As Integer = 0
        Do While (i <= recordFields)
            Try
                If IsDBNull(originalRecord(i)) Then
                    If Not IsDBNull(ModifiedRecord.Item(i)) Then
                        builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, "NULL", ModifiedRecord.Item(i)))
                    End If
                ElseIf Information.IsDBNull(ModifiedRecord.Item(i)) Then
                    If Not Information.IsDBNull(originalRecord(i)) Then
                        builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, originalRecord(i), "NULL"))
                    End If
                ElseIf ModifiedRecord.Item(i) IsNot originalRecord(i) Then
                    builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, originalRecord(i), ModifiedRecord.Item(i)))
                End If
            Catch
            End Try
            i += 1
        Loop
        If (builder.Length > 2) Then
            builder.Length = (builder.Length - 2)
            AuditTrailBLL.AddTrailEntry(My.Settings.UserName, CType(ModifiedRecord.Item(0), Integer), "Shifts", builder.ToString)
        End If

    End Sub

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Select, True)> _
    Public Function GetShifts() As SPG.ShiftsDataTable

        Return Adapter.GetShifts()

    End Function

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Update, True)> _
    Public Function UpdateShift(ByVal id As Nullable(Of Integer), ByVal name As String) As Boolean

        Dim shifts As SPG.ShiftsDataTable = Adapter.GetShiftByID(id.Value)

        If shifts.Count = 0 Then
            'It is a new Product
            Return Me.InsertShift(name)
        End If

        Dim shift As SPG.ShiftsRow = shifts(0)

        If String.IsNullOrEmpty(name) Then
            Throw New ApplicationException("You must provide a Shift Name.")
        End If

        Dim originalRecord As Object() = Nothing
        originalRecord = shift.ItemArray

        shift.ShiftName = name

        If Not IsNothing(originalRecord) Then
            Me.UpdateAuditTrail(shift, originalRecord)
        End If

        Dim rowsAffected As Integer = Adapter.Update(shift)

        Return rowsAffected = 1

    End Function

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Insert, True)> _
    Public Function InsertShift(ByVal name As String) As Boolean

        Dim shifts As SPG.ShiftsDataTable = New SPG.ShiftsDataTable
        Dim shift As SPG.ShiftsRow = shifts.NewShiftsRow

        If String.IsNullOrEmpty(name) Then
            Throw New ApplicationException("You must provide a Shift Name.")
        End If

        shift.ShiftName = name

        shifts.AddShiftsRow(shift)
        Dim rowsAffected As Integer = Adapter.Update(shifts)

        Return rowsAffected = 1

    End Function

    <System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Delete, True)> _
    Public Function DeleteShift(ByVal id As Integer) As Boolean

        Dim shifts As SPG.ShiftsDataTable = Adapter.GetShiftByID(id)
        Dim rowsAffected As Integer = 0

        If shifts.Count = 1 Then
            Dim shift As SPG.ShiftsRow = CType(shifts.Rows(0), SPG.ShiftsRow)
            If New ProductionBLL().GetProductionByShift(id).Count <> 0 Then
                MessageBox.Show("You can't delete this shift there are other records associated to it")
                Return True
            End If
            rowsAffected = Adapter.Delete(id, shift.ts)
        End If

        'Return true if precisely one row was deleted, otherwise return false.
        Return rowsAffected = 1

    End Function

    Public Shared Function GetShiftByID(ByVal shiftID As Integer) As Shifts

        Return XpoDefault.Session.GetObjectByKey(Of Shifts)(shiftID)

    End Function

End Class

File: SuperiorPackGroup/BLL/ShiftsBLL.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;

using SuperiorPackGroup.SPGTableAdapters;
using System.Text;
using DXDAL.SPGData;
using DevExpress.Xpo;
using System.Linq;

namespace SuperiorPackGroup
{
	[System.ComponentModel.DataObject()]
	public class ShiftsBLL
	{

		private ShiftsTableAdapter m_ShiftsTableAdapter = null;
		public ShiftsTableAdapter Adapter
		{

			get
			{
				if (m_ShiftsTableAdapter == null)
				{
					m_ShiftsTableAdapter = new ShiftsTableAdapter();
					m_ShiftsTableAdapter.Connection.ConnectionString = Properties.Settings.Default.UserConnectionString;
				}
				return m_ShiftsTableAdapter;
			}

		}

		private void UpdateAuditTrail(SPG.ShiftsRow ModifiedRecord, object[] originalRecord)
		{

			StringBuilder builder = new StringBuilder(string.Empty);
			int recordFields = (ModifiedRecord.ItemArray.Length - 1);
			int i = 0;
			while (i <= recordFields)
			{
				try
				{
					if (Convert.IsDBNull(originalRecord[i]))
					{
						if (!Convert.IsDBNull(ModifiedRecord[i]))
						{
							builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, "NULL", ModifiedRecord[i]));
						}
					}
					else if (Convert.IsDBNull(ModifiedRecord[i]))
					{
						if (!Convert.IsDBNull(originalRecord[i]))
						{
							builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, originalRecord[i], "NULL"));
						}
					}
					else if (ModifiedRecord[i] != originalRecord[i])
					{
						builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, originalRecord[i], ModifiedRecord[i]));
					}
				}
				catch
				{
				}
				i += 1;
			}
			if (builder.Length > 2)
			{
				builder.Length = (builder.Length - 2);
				AuditTrailBLL.AddTrailEntry(Properties.Settings.Default.UserName, Convert.ToInt32(ModifiedRecord[0]), "Shifts", builder.ToString());
			}

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Select, true)]
		public SPG.ShiftsDataTable GetShifts()
		{

			return Adapter.GetShifts();

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Update, true)]
		public bool UpdateShift(int? id, string name)
		{

			SPG.ShiftsDataTable shifts = Adapter.GetShiftByID(id.Value);

			if (shifts.Count() == 0)
			{
				//It is a new Product
				return this.InsertShift(name);
			}

			SPG.ShiftsRow shift = shifts[0];

			if (string.IsNullOrEmpty(name))
			{
				throw new ApplicationException("You must provide a Shift Name.");
			}

			object[] originalRecord = shift.ItemArray;

			shift.ShiftName = name;

			if (!(originalRecord == null))
			{
				this.UpdateAuditTrail(shift, originalRecord);
			}

			int rowsAffected = Adapter.Update(shift);

			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Insert, true)]
		public bool InsertShift(string name)
		{

			SPG.ShiftsDataTable shifts = new SPG.ShiftsDataTable();
			SPG.ShiftsRow shift = shifts.NewShiftsRow();

			if (string.IsNullOrEmpty(name))
			{
				throw new ApplicationException("You must provide a Shift Name.");
			}

			shift.ShiftName = name;

			shifts.AddShiftsRow(shift);
			int rowsAffected = Adapter.Update(shifts);

			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Delete, true)]
		public bool DeleteShift(int id)
		{

			SPG.ShiftsDataTable shifts = Adapter.GetShiftByID(id);
			int rowsAffected = 0;

			if (shifts.Count() == 1)
			{
				SPG.ShiftsRow shift = (SPG.ShiftsRow)shifts.Rows[0];
				if ((new ProductionBLL()).GetProductionByShift(id).Count != 0)
				{
					MessageBox.Show("You can't delete this shift there are other records associated to it");
					return true;
				}
				rowsAffected = Adapter.Delete(id, shift.ts);
			}

			//Return true if precisely one row was deleted, otherwise return false.
			return rowsAffected == 1;

		}

		public static Shifts GetShiftByID(int shiftID)
		{

			return XpoDefault.Session.GetObjectByKey<Shifts>(shiftID);

		}

	}

}